src/knx/security_interface_object.cpp

```diff
@@ -0,0 +1,197 @@
+                uint8_t info = data[2];
+
+                // failure counters
+                if (id == 0 && info == 0)
+                {
+                    resultData[0] = ReturnCodes::Success;
+                    resultData[1] = id;
+                    resultData[2] = info;
+                    obj->_secAppLayer->getFailureCounters(&resultData[3]); // Put 8 bytes in the buffer
+                    resultLength = 3 + 8;
+                    return;
+                }
+                // query latest failure by index
+                else if(id == 1)
+                {
+                    uint8_t maxBufferSize = resultLength; // Remember the maximum buffer size of the buffer that is provided to us
+                    uint8_t index = info;
+                    uint8_t numBytes = obj->_secAppLayer->getFromFailureLogByIndex(index, &resultData[2], maxBufferSize);
+                    if ( numBytes > 0)
+                    {
+                        resultData[0] = ReturnCodes::Success;
+                        resultData[1] = id;
+                        resultData[2] = index;
+                        resultLength += numBytes;
+                        resultLength = 3 + numBytes;
+                        return;
+                    }
+                    resultData[0] = ReturnCodes::DataVoid;
+                    resultData[1] = id;
+                    resultLength = 2;
+                    return;
+                }
+                resultData[0] = ReturnCodes::GenericError;
+                resultLength = 1;
+            }),
+        new DataProperty( PID_TOOL_KEY, true, PDT_GENERIC_16, 1, ReadLv3 | WriteLv0, (uint8_t*) _fdsk ), // default is FDSK // TODO: do not overwrite on every device startup!!!!!!!!! ETS changes this property during programming from FDSK to some random key!
+        new DataProperty( PID_SECURITY_REPORT, true, PDT_BITSET8, 1, ReadLv3 | WriteLv0, _secReport ), // Not implemented
+        new DataProperty( PID_SECURITY_REPORT_CONTROL, true, PDT_BINARY_INFORMATION, 1, ReadLv3 | WriteLv0, _secReportCtrl ), // Not implemented
+        new DataProperty( PID_SEQUENCE_NUMBER_SENDING, true, PDT_GENERIC_06, 1, ReadLv3 | WriteLv0 ), // Updated by our device accordingly
+        new DataProperty( PID_ZONE_KEY_TABLE, true, PDT_GENERIC_19, 32, ReadLv3 | WriteLv0 ), // written by ETS
+        new DataProperty( PID_GO_SECURITY_FLAGS, true, PDT_GENERIC_01, 32, ReadLv3 | WriteLv0 ), // written by ETS
+        new DataProperty( PID_ROLE_TABLE, true, PDT_GENERIC_01, 32, ReadLv3 | WriteLv0 ), // written by ETS
+        new DataProperty( PID_TOOL_SEQUENCE_NUMBER_SENDING, true, PDT_GENERIC_06, 1, ReadLv3 | WriteLv0 ) // Updated by our device accordingly (non-standardized!)
+    };
+    initializeProperties(sizeof(properties), properties);
+}
+
+void SecurityInterfaceObject::secureApplicationLayer(SecureApplicationLayer& secAppLayer)
+{
+    _secAppLayer = &secAppLayer;
+}
+
+uint8_t* SecurityInterfaceObject::save(uint8_t* buffer)
```

Comment:
You can just drop the save, restore and saveSize methods.

Reply:
Yes, I will remove it if I clean up everything.

Reply:
done.